##Tugas 2 Praktikum Pembelajaran Mesin Lab A1

###Nama : Bonifasius Bryan Suryaningtyas
###NIM  : 24060121130066

###Soal 1 - Eksplorasi *Dataset* Iris dengan Model Klasifikasi di Luar Modul

In [1]:
import pandas as pd

##### Penjelasan *Source Code* :
Digunakan untuk meng-*import* modul pandas ke dalam kode dengan alias "pd", memungkinkan kita untuk menggunakan berbagai fungsi yang disediakan oleh pandas dalam hal mengelola dan menganalisis data.


In [2]:
url = "http://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

nama_kolom = ['sepal-l', 'sepal-w', 'petal-l', 'petal-w', 'class']

dataset = pd.read_csv(url, names=nama_kolom)

##### Penjelasan *Source Code* :
1. "url = "http://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"" digunakan untuk menyimpan URL file CSV ke dalam variabel bernama url.
2. "nama_kolom = ['sepal-l', 'sepal-w', 'petal-l', 'petal-w', 'class']"
digunakan untuk mendefinisikan daftar nama kolom.
3. "dataset = pd.read_csv(url, names=nama_kolom)" digunakan untuk membaca *file* CSV dari URL dan memberikan nama kolom sesuai daftar yang telah dibuat.

In [3]:
dataset.head(5)

sepal-l  sepal-w  petal-l  petal-w        class
0      5.1      3.5      1.4      0.2  Iris-setosa
1      4.9      3.0      1.4      0.2  Iris-setosa
2      4.7      3.2      1.3      0.2  Iris-setosa
3      4.6      3.1      1.5      0.2  Iris-setosa
4      5.0      3.6      1.4      0.2  Iris-setosa

##### Penjelasan *Source Code* :
Digunakan untuk menampilkan *record* lima data teratas pada *dataset*.

In [4]:
dataset.shape

(150, 5)

##### Penjelasan *Source Code* :
Digunakan untuk mencetak dimensi ukuran *dataset*. Hasil dari "dataset.shape" adalah (150, 5), yang berarti DataFrame *dataset* memiliki 150 baris dan 5 kolom.

In [22]:
from sklearn import model_selection

array = dataset.values

X = array[:,0:4]
Y = array[:,4]

validation_size = 0.20

seed = 7

X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

##### Penjelasan *Source Code* :
1. "from sklearn import model_selection" digunakan untuk meng-*impor* modul model_selection dari library scikit-learn (sklearn). Modul ini menyediakan berbagai fungsi yang berguna untuk melakukan seleksi model, evaluasi model, dan pemisahan *dataset* menjadi bagian pelatihan dan pengujian.
2. "array = dataset.values" digunakan untuk mengonversi *dataset* ke dalam bentuk array menggunakan metode values.
3. "X = array[:,0:4]" digunakan untuk melakukan indeksing untuk memilih semua baris (:) dan kolom dari 0 hingga 3 (0:4). Ini berarti kita mengambil kolom-kolom 0, 1, 2, dan 3 sebagai fitur untuk variabel X.
4. "Y = array[:,4]" digunakan untuk melakukan indeksing untuk memilih semua baris (:) dan hanya kolom ke-4 (4). Ini berarti kita mengambil kolom 4 sebagai target atau label untuk variabel Y.
5. "validation_size = 0.20" digunakan untuk menetapkan ukuran *data validasi* sebagai 20% dari keseluruhan *dataset*.
6. "seed = 7" digunakan untuk memastikan hasil pemisahan *dataset* yang konsisten ketika kode dijalankan berulang kali. Berperan dalam pengacakan elemen.
7. "X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)" digunakan untuk membagi *dataset* menjadi dua bagian, yaitu *data training* dan *data validation*.

In [12]:
print("X_train : ", X_train.shape)
print("X_validation : ", X_validation.shape)
print("Y_train : ", Y_train.shape)
print("Y_validation : ", Y_validation.shape)

X_train :  (120, 4)
X_validation :  (30, 4)
Y_train :  (120,)
Y_validation :  (30,)


##### Penjelasan *Source Code* dan Hasil :
Digunakan untuk menampilkan jumlah data yang masuk dalam kategori *data training* dan *data validation*. Sebagai contoh, "X_train :  (120, 4)" mengindikasikan bahwa terdapat 120 data yang berperan sebagai *data training* dengan masing-masing 4 atribut tiap datanya. Atau "Y_validation : (30,)" yang menunjukkan terdapat 30 data target yang berperan sebagai *data validation*.

In [13]:
seed = 7
scoring = 'accuracy'

##### Penjelasan *Source Code* :
1. "seed = 7" digunakan untuk memastikan hasil pemisahan *dataset* yang konsisten ketika kode dijalankan berulang kali. Berperan dalam pengacakan elemen.
2. "scoring = 'accuracy'" digunakan untuk menentukan metrik evaluasi yang akan digunakan saat melakukan evaluasi model. Dalam hal ini, metrik evaluasi yang dipilih adalah akurasi.

In [35]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import RandomForestClassifier

##### Penjelasan *Source Code* :
Digunakan untuk meng-*impor* *library* yang dibutuhkan oleh model dalam melakukan tugas evaluasi algoritma. Disini saya menggunakan tiga model, yaitu Decision Trees, AdaBoost, dan Random Forest.

In [36]:
models = []

models.append(('Decision Trees', DecisionTreeClassifier()))
models.append(('AdaBoost', AdaBoostClassifier()))
models.append(('Random Forest', RandomForestClassifier()))

##### Penjelasan *Source Code* :
1. "models = []" digunakan untuk membuat *list* kosong yang disebut models. *List* ini akan digunakan untuk menyimpan informasi nama model dan objek model yang telah diinisialisasi.
2. Tiga *line code* sisanya digunakan untuk membuat kumpulan model yang berbeda dengan konfigurasi yang berbeda pula, sehingga kita dapat membandingkan kinerja mereka pada *dataset* tertentu.

In [37]:
results = []
names = []

for name, model in models:
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)
  results.append(cv_results)
  names.append(name)
  msg = "%s : %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

Decision Trees : 0.950000 (0.076376)
AdaBoost : 0.966667 (0.040825)
Random Forest : 0.966667 (0.040825)


##### Penjelasan *Source Code* :
1. "results = []" digunakan untuk membuat *list* kosong yang akan berguna untuk menyimpan hasil evaluasi dari model-model yang diuji. Setiap model akan memiliki sejumlah *cross-validation scores*, dan *list* ini akan menyimpan skor-skor tersebut untuk setiap model.
2. "names = []" digunakan untuk membuat *list* kosong yang akan berguna untuk menyimpan nama-nama model yang diuji. Setiap model akan memiliki nama yang diberikan pada saat pengujian, dan *list* ini menyimpan nama-nama tersebut.
3. Bagian *looping* berjalan melalui setiap model yang ada dalam *list* models. Setiap iterasi *loop* mengikuti urutan langkah yaitu **a.** Membuat objek kFold untuk melaksanakan *cross-validation* dengan model_selection.KFold, **b.** Menghitung hasil *cross-validation scores* (cv_results) menggunakan model_selection.cross_val_score. Skor ini akan mencakup *cross-validation scores* dari tiap lipatan (*fold*), **c.** Menambahkan skor-skor tersebut ke dalam *list* results, **d.** Menambahkan nama model ke dalam *list* names, dan **e.** Mencetak hasil akhir yang berisi nama model, rata-rata skor, dan standar deviasi dari hasil *cross-validation scores*.

##### Penjelasan Hasil :
1. "Decision Trees : 0.950000 (0.076376)" menunjukkan bahwa model Decision Trees memiliki akurasi rata-rata sekitar 95% dan standar deviasi sekitar 7.64%.
2. "AdaBoost : 0.966667 (0.040825)" menunjukkan bahwa model AdaBoost memiliki akurasi rata-rata sekitar 96.67%, yang sedikit lebih tinggi dari Decision Trees, dan standar deviasi sekitar 4.08%.
3. "Random Forest : 0.966667 (0.040825)" menunjukkan bahwa model Random Forest memberikan hasil yang serupa dengan AdaBoost, dengan akurasi rata-rata sekitar 96.67% dan standar deviasi sekitar 4.08%. Ini menunjukkan bahwa kedua model ini memiliki performa yang serupa pada *dataset* ini.

In [38]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

##### Penjelasan *Source Code* :
Digunakan untuk meng-*impor* *library* metrik evaluasi seperti accuracy_score, confusion_matrix, dan classification_report. Gunanya untuk mengukur dan menganalisis kinerja model klasifikasi tadi.

In [41]:
random_forest = RandomForestClassifier()
random_forest.fit(X_train, Y_train)
random_forest_predictions = random_forest.predict(X_validation)

print(accuracy_score(Y_validation, random_forest_predictions))
print(confusion_matrix(Y_validation, random_forest_predictions))
print(classification_report(Y_validation, random_forest_predictions))

0.9
[[ 7  0  0]
 [ 0 10  2]
 [ 0  1 10]]
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.91      0.83      0.87        12
 Iris-virginica       0.83      0.91      0.87        11

       accuracy                           0.90        30
      macro avg       0.91      0.91      0.91        30
   weighted avg       0.90      0.90      0.90        30



##### Penjelasan *Source Code* :
1. "random_forest = RandomForestClassifier()" digunakan untuk membuat objek model RandomForestClassifier yang akan digunakan dalam analisis dan prediksi.
2. "random_forest.fit(X_train, Y_train)" digunakan untuk melatih model RandomForestClassifier dengan menggunakan *data training* X_train dan label Y_train, sehingga model dapat memahami pola dalam data.
3. "random_forest_predictions = random_forest.predict(X_validation)" digunakan untuk membuat prediksi dengan model RandomForestClassifier pada *data validation* X_validation. Hasil prediksi kemudian disimpan dalam variabel random_forest_predictions.
4. "print(accuracy_score(Y_validation, random_forest_predictions))" digunakan untuk mencetak akurasi model, memberikan informasi tentang seberapa baik model dapat memprediksi dengan benar kelas dari *data validation*.
5. "print(confusion_matrix(Y_validation, random_forest_predictions))" digunakan untuk mencetak *confusion matrix*, yang memberikan gambaran tentang seberapa baik model dapat membedakan antara kelas-kelas yang berbeda pada *data validation*.
6. "print(classification_report(Y_validation, random_forest_predictions))" digunakan untuk mencetak laporan klasifikasi, yang memberikan informasi terperinci tentang metrik evaluasi seperti presisi, recall, dan f1-score untuk setiap kelas pada *data validation*.

##### Penjelasan Hasil :
1. Hasil dari *print* pertama adalah accuracy_score. Akurasi model RandomForestClassifier adalah sekitar 86.67%. Ini menunjukkan bahwa model berhasil memprediksi kelas dengan benar untuk sekitar 86.67% dari total *data validation*.
2. Hasil dari *print* kedua adalah confusion_matrix. *Confusion Matrix* menunjukkan bahwa terdapat 7 prediksi benar untuk kelas "Iris-setosa", 10 prediksi benar untuk kelas "Iris-versicolor", dan 9 prediksi benar untuk kelas "Iris-virginica". Terdapat juga beberapa prediksi yang salah.
3. Hasil dari *print* ketiga adalah classification_report. **Pertama**, *precision* menunjukkan kemampuan model dalam memberikan prediksi yang benar untuk suatu kelas. Tingkat *precision* tinggi menunjukkan sedikitnya prediksi *false positive*. Misalnya, *precision* untuk "Iris-setosa" adalah 100%, yang artinya tidak ada prediksi *false positive* untuk kelas ini. **Kedua**, *recall* yang menunjukkan *positive instance proportion* yang benar diprediksi oleh model. *Recall* yang tinggi menunjukkan bahwa model dapat menangkap sebagian besar *positive instance*. Misalnya, *recall* untuk "Iris-versicolor" adalah 83%, yang artinya model berhasil menangkap 83% dari semua *instance* "Iris-versicolor". **Ketiga**, *F1-Score* menunjukkan rata-rata antara *precision* dan *recall*. *F1-Score* yang tinggi menunjukkan keseimbangan antara *precision* dan *recall*. Misalnya, *F1-Score* untuk "Iris-setosa" adalah 1.00, menunjukkan keseimbangan yang baik antara *precision* dan *recall*. **Keempat**, support menunjukkan jumlah *instance* dalam setiap kelas.

###Soal 2 - Eksplorasi *Dataset* Ecoli dengan Model KNN, NB, dan SVM

In [42]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/ecoli/ecoli.data"

nama_kolom = ["Sequence Name", "mcg", "gvh", "lip", "chg", "aac", "alm1", "alm2", "Class"]

dataset = pd.read_csv(url, names=nama_kolom, header=None, sep='\\s+')

##### Penjelasan *Source Code* :
1. "url = "http://archive.ics.uci.edu/ml/machine-learning-databases/ecoli/ecoli.data"" digunakan untuk menyimpan URL *file* CSV ke dalam variabel bernama url.
2. "nama_kolom = ["Sequence Name", "mcg", "gvh", "lip", "chg", "aac", "alm1", "alm2", "Class"]" digunakan untuk mendefinisikan daftar nama kolom.
3. "dataset = pd.read_csv(url, names=nama_kolom, header=None, sep='\\s+')" digunakan untuk membaca *file* dari URL, memberikan nama kolom sesuai daftar yang telah dibuat, mengabaikan *header*, dan menggunakan spasi sebagai pemisah antar kolom.

In [43]:
dataset.head(5)

Sequence Name   mcg   gvh   lip  chg   aac  alm1  alm2 Class
0     AAT_ECOLI  0.49  0.29  0.48  0.5  0.56  0.24  0.35    cp
1    ACEA_ECOLI  0.07  0.40  0.48  0.5  0.54  0.35  0.44    cp
2    ACEK_ECOLI  0.56  0.40  0.48  0.5  0.49  0.37  0.46    cp
3    ACKA_ECOLI  0.59  0.49  0.48  0.5  0.52  0.45  0.36    cp
4     ADI_ECOLI  0.23  0.32  0.48  0.5  0.55  0.25  0.35    cp

##### Penjelasan *Source Code* :
Digunakan untuk menampilkan *record* lima data teratas pada *dataset*.

In [44]:
dataset.shape

(336, 9)

##### Penjelasan *Source Code* :
Digunakan untuk mencetak dimensi ukuran *dataset*. Hasil dari "dataset.shape" adalah (336, 9), yang berarti DataFrame *dataset* memiliki 336 baris dan 9 kolom.

In [45]:
from sklearn import model_selection

array = dataset.values

X = array[:,1:8]
Y = array[:,8]

validation_size = 0.20

seed = 7

X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

##### Penjelasan *Source Code* :
1. "from sklearn import model_selection" digunakan untuk meng-*impor* modul model_selection dari library scikit-learn (sklearn). Modul ini menyediakan berbagai fungsi yang berguna untuk melakukan seleksi model, evaluasi model, dan pemisahan *dataset* menjadi bagian pelatihan dan pengujian.
2. "array = dataset.values" digunakan untuk mengonversi *dataset* ke dalam bentuk array menggunakan metode values.
3. "X = array[:,1:8]" digunakan untuk melakukan indeksing untuk memilih semua baris (:) dan kolom dari 1 hingga 7 (1:8). Ini berarti kita mengambil kolom-kolom 1, 2, 3, 4, 5, 6, dan 7 sebagai fitur untuk variabel X.
4. "Y = array[:,8]" digunakan untuk melakukan indeksing untuk memilih semua baris (:) dan hanya kolom ke-8 (8). Ini berarti kita mengambil kolom 8 sebagai target atau label untuk variabel Y.
5. "validation_size = 0.20" digunakan untuk menetapkan ukuran *data validation* sebagai 20% dari keseluruhan *dataset*.
6. "seed = 7" digunakan untuk memastikan hasil pemisahan *dataset* yang konsisten ketika kode dijalankan berulang kali. Berperan dalam pengacakan elemen.
7. "X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)" digunakan untuk membagi *dataset* menjadi dua bagian, yaitu *data training* dan *data validation*.

In [47]:
print("X_train : ", X_train.shape)
print("X_validation : ", X_validation.shape)
print("Y_train : ", Y_train.shape)
print("Y_validation : ", Y_validation.shape)

X_train :  (268, 7)
X_validation :  (68, 7)
Y_train :  (268,)
Y_validation :  (68,)


##### Penjelasan *Source Code* dan Hasil :
Digunakan untuk menampilkan jumlah data yang masuk dalam kategori *data training* dan *data validation*. Sebagai contoh, "X_train :  (268, 7)" mengindikasikan bahwa terdapat 268 data yang berperan sebagai *data training* dengan masing-masing 7 atribut tiap datanya. Atau "Y_validation : (68,)" yang menunjukkan terdapat 68 data target yang berperan sebagai *data validation*.

In [48]:
seed = 7
scoring = 'accuracy'

##### Penjelasan *Source Code* :
1. "seed = 7" digunakan untuk memastikan hasil pemisahan *dataset* yang konsisten ketika kode dijalankan berulang kali. Berperan dalam pengacakan elemen.
2. "scoring = 'accuracy'" digunakan untuk menentukan metrik evaluasi yang akan digunakan saat melakukan evaluasi model. Dalam hal ini, metrik evaluasi yang dipilih adalah akurasi.

In [50]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC

##### Penjelasan *Source Code* :
Digunakan untuk meng-*impor* *library* yang dibutuhkan oleh model dalam melakukan tugas evaluasi algoritma. Disini digunakan tiga model, yaitu K-Neighbors Neighbors, Gaussian Naive Bayes, dan Support Vector Machines.

In [53]:
models = []

models.append(('K-Neighbors Neighbors', KNeighborsClassifier()))
models.append(('Gaussian Naive Bayes', GaussianNB()))
models.append(('Support Vector Machines', SVC()))

##### Penjelasan *Source Code* :
1. "models = []" digunakan untuk membuat *list* kosong yang disebut models. *List* ini akan digunakan untuk menyimpan informasi nama model dan objek model yang telah diinisialisasi.
2. Tiga *line code* sisanya digunakan untuk membuat kumpulan model yang berbeda dengan konfigurasi yang berbeda pula, sehingga kita dapat membandingkan kinerja mereka pada *dataset* tertentu.

In [54]:
results = []
names = []

for name, model in models:
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)
  results.append(cv_results)
  names.append(name)
  msg = "%s : %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

K-Neighbors Neighbors : 0.854558 (0.090490)
Gaussian Naive Bayes : 0.742877 (0.106148)
Support Vector Machines : 0.880342 (0.062807)


##### Penjelasan *Source Code* :
1. "results = []" digunakan untuk membuat *list* kosong yang akan berguna untuk menyimpan hasil evaluasi dari model-model yang diuji. Setiap model akan memiliki sejumlah *cross-validation scores*, dan *list* ini akan menyimpan skor-skor tersebut untuk setiap model.
2. "names = []" digunakan untuk membuat *list* kosong yang akan berguna untuk menyimpan nama-nama model yang diuji. Setiap model akan memiliki nama yang diberikan pada saat pengujian, dan *list* ini menyimpan nama-nama tersebut.
3. Bagian *looping* berjalan melalui setiap model yang ada dalam *list* models. Setiap iterasi *loop* mengikuti urutan langkah yaitu **a.** Membuat objek kFold untuk melaksanakan *cross-validation* dengan model_selection.KFold, **b.** Menghitung hasil *cross-validation scores* (cv_results) menggunakan model_selection.cross_val_score. Skor ini akan mencakup *cross-validation scores* dari tiap lipatan (*fold*), **c.** Menambahkan skor-skor tersebut ke dalam *list* results, **d.** Menambahkan nama model ke dalam *list* names, dan **e.** Mencetak hasil akhir yang berisi nama model, rata-rata skor, dan standar deviasi dari hasil *cross-validation scores*.

##### Penjelasan Hasil :
1. "K-Neighbors Neighbors : 0.854558 (0.090490)" menunjukkan bahwa model K-Neighbors Neighbors memiliki akurasi rata-rata sekitar 85.5% dan standar deviasi sekitar 9.05%.
2. "Gaussian Naive Bayes : 0.742877 (0.106148)" menunjukkan bahwa model Gaussian Naive Bayes memiliki akurasi rata-rata sekitar 74.3% dan standar deviasi sekitar 10.62%.
3. "Support Vector Machines : 0.880342 (0.062807)" menunjukkan bahwa model Support Vector Machines memiliki akurasi rata-rata sekitar 88.3% dan standar deviasi sekitar 6.28%.

In [55]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

##### Penjelasan *Source Code* :
Digunakan untuk meng-*impor* *library* metrik evaluasi seperti accuracy_score, confusion_matrix, dan classification_report. Gunanya untuk mengukur dan menganalisis kinerja model klasifikasi tadi.

In [57]:
svm = SVC()
svm.fit(X_train, Y_train)
predictions = svm.predict(X_validation)

print(accuracy_score(Y_validation, predictions))
print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))

0.8529411764705882
[[28  0  0  0  0  2]
 [ 0 10  2  0  0  0]
 [ 0  4  4  0  0  0]
 [ 0  0  0  2  0  1]
 [ 0  0  0  0  2  0]
 [ 1  0  0  0  0 12]]
              precision    recall  f1-score   support

          cp       0.97      0.93      0.95        30
          im       0.71      0.83      0.77        12
         imU       0.67      0.50      0.57         8
          om       1.00      0.67      0.80         3
         omL       1.00      1.00      1.00         2
          pp       0.80      0.92      0.86        13

    accuracy                           0.85        68
   macro avg       0.86      0.81      0.82        68
weighted avg       0.86      0.85      0.85        68



##### Penjelasan *Source Code* :
1. "svm = SVC()" digunakan untuk membuat objek model SVC yang akan digunakan dalam analisis dan prediksi.
2. "svm.fit(X_train, Y_train)" digunakan untuk melatih model SVC dengan menggunakan *data training* X_train dan label Y_train, sehingga model dapat memahami pola dalam data.
3. "predictions = svm.predict(X_validation)" digunakan untuk membuat prediksi dengan model SVC pada *data validation* X_validation. Hasil prediksi kemudian disimpan dalam variabel predictions.
4. "print(accuracy_score(Y_validation, predictions))" digunakan untuk mencetak akurasi model, memberikan informasi tentang seberapa baik model dapat memprediksi dengan benar kelas dari *data validation*.
5. "print(confusion_matrix(Y_validation, predictions))" digunakan untuk mencetak *confusion matrix*, yang memberikan gambaran tentang seberapa baik model dapat membedakan antara kelas-kelas yang berbeda pada *data validation*.
6. "print(classification_report(Y_validation, predictions))" digunakan untuk mencetak laporan klasifikasi, yang memberikan informasi terperinci tentang metrik evaluasi seperti presisi, recall, dan f1-score untuk setiap kelas pada *data validation*.

##### Penjelasan Hasil :
1. Hasil dari *print* pertama adalah accuracy_score. Akurasi model Support Vector Machine (SVM) untuk klasifikasi *dataset* adalah sekitar 85.29%. Ini mengindikasikan bahwa model dengan parameter yang digunakan berhasil memprediksi kelas-kelas pada *data validation* dengan akurasi sekitar 85.29%.
2. Hasil dari *print* kedua adalah confusion_matrix. *Confusion Matrix* menunjukkan bahwa terdapat 28 prediksi benar untuk kelas "cp", 10 prediksi benar untuk kelas "im", 4 prediksi benar untuk kelas "imU", 2 prediksi benar untuk kelas "om", 2 prediksi benar untuk kelas "omL", dan 12 prediksi benar untuk kelas "pp".
3. Hasil dari *print* ketiga adalah classification_report. **Pertama**, *precision* menunjukkan kemampuan model dalam memberikan prediksi yang benar untuk suatu kelas. Tingkat *precision* tinggi menunjukkan sedikitnya prediksi *false positive*. Misalnya, *precision* untuk kelas "cp" adalah 0.97, yang artinya sebagian besar prediksi positif untuk kelas "cp" benar. **Kedua**, *recall* yang menunjukkan *positive instance proportion* yang benar diprediksi oleh model. *Recall* yang tinggi menunjukkan bahwa model dapat menangkap sebagian besar *positive instance*. Misalnya, *recall* untuk kelas "im" adalah 0.83, yang artinya model berhasil menangkap 83% dari semua *instance* "im". **Ketiga**, *F1-Score* menunjukkan rata-rata antara *precision* dan *recall*. *F1-Score* yang tinggi menunjukkan keseimbangan antara *precision* dan *recall*. Misalnya, *F1-Score* untuk "omL" adalah 1.00, menunjukkan keseimbangan yang baik antara *precision* dan *recall*. **Keempat**, support menunjukkan jumlah *instance* dalam setiap kelas.

Link Collab : https://colab.research.google.com/drive/1ZcYfVylv5swZaWOjP0LeBROEk8YdaKSc?authuser=0#scrollTo=BWR7K4FORkpP

###Terima Kasih